src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala

```diff
@@ -20,11 +20,26 @@ trait DataHelpers extends QueryStringListsServer with Validation with akkahttp.s
   import io.circe.syntax._
 
   /** Function validating request for the query endpoint */
-  override def validated[A](response: A => Route, invalidDocs: Documentation): Either[List[QueryValidationError], A] => Route = {
+  override def validated(response: QueryResponseWithOutput => Route, invalidDocs: Documentation): Either[List[QueryValidationError], QueryResponseWithOutput] => Route = {
     case Left(errors) =>
       complete(StatusCodes.BadRequest -> s"Errors: \n${errors.mkString("\n")}")
-    case Right(success) =>
+    case Right(success@QueryResponseWithOutput(_, OutputType.json)) =>
       response(success)
+    case Right(QueryResponseWithOutput(queryResponse, OutputType.csv)) =>
+      complete(StatusCodes.OK -> makeCsv(queryResponse))
+  }
+
+  def makeCsv(queryResponse: List[QueryResponse]): String = {
```

Comment:
Idea: what about re-using the Conversion type class we have to handout the work of transforming the map to  csv in a separate, independent place and context?

Reply:
In any case, I would try personally to make some effort to define the csv encoding separately instead of "inlining" that here in the route code

Reply:
you are absolutely right, I haven't thought of that. Implemented your way.